Ethernet: Add stm32n6 support

marwaiehm-st opened the pull request:
Add STM32N6 ethernet support:

- Add the compatible of the STM32N6 series
- Update STM32N6 pinctrl dtsi files with the ETH pins
- Align Ethernet descriptors to 32 bytes for STM32N6 to ensure efficient DMA operations and improve cache line efficiency
 and overall performance
- Update the ETH_STM32_HAL menu configuration to conditionally select USE_STM32_HAL_RIF if SOC_SERIES_STM32N6X is enabled.
- Add RISAF configuration in eth_initialize function for STM32N6 series to set up master and slave security attributes
for the Ethernet peripheral.
- Ensure RISAF configuration is done before enabling the Ethernet clock to maintain proper security attributes.

This PR supports only the Nucleo_N657X0_Q. I will add support for the STM32N6570_DK in a separate PR.

marwaiehm-st committed:
dts: arm: st: n6: Add the ethernet node

Add the ethernet node

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
boards: st: nucleo_n657x0_q: Add ethernet node

Add ethernet node

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
west.yml: hal_stm32: Add stm32n6 ethernet pinctrl

Update STM32N6 pinctrl dtsi files with the ETH pins
Add STM32N6 Ethernet pinctrl config to stm32-pinctrl-config.yaml

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -79,6 +80,10 @@ static const struct device *eth_stm32_phy_dev = DEVICE_PHY_BY_NAME(0);
 #define ETH_MEDIA_INTERFACE_MII		HAL_ETH_MII_MODE
 #define ETH_MEDIA_INTERFACE_RMII	HAL_ETH_RMII_MODE
 
+#if defined(CONFIG_SOC_SERIES_STM32N6X)
```
Let's add a dedicated compatible

marwaiehm-st replied:
Done

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -419,6 +430,14 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
 	/* Reset TX complete interrupt semaphore before TX request*/
 	k_sem_reset(&dev_data->tx_int_sem);
 
+	/**
+	 * Assure cache coherency before DMA read operation.
+	 * Clean the data cache for the memory region specified by tx_config.TxBuffer->buffer
+	 * and tx_config.TxBuffer->len to ensure that any modified data in the cache is written
+	 * back to the main memory before the DMA controller reads it.
+	 */
+	sys_cache_data_flush_range((void *)(tx_config.TxBuffer->buffer), tx_config.TxBuffer->len);
+
```
Please put in a dedicated commit

marwaiehm-st replied:
Done

marwaiehm-st committed:
drivers: ethernet: eth_stm32_hal: Add stm32n6 ethernet support

-Update the ETH_STM32_HAL menu configuration to conditionally
 select USE_STM32_HAL_RIF if SOC_SERIES_STM32N6X is enabled.

-Align Ethernet descriptors to 32 bytes for STM32N6 to ensure
 efficient DMA operations and improve cache line efficiency
 and overall performance

-Add RISAF configuration in eth_initialize function for STM32N6
 series to set up master and slave security attributes
 for the Ethernet peripheral.

-Ensure RISAF configuration is done before enabling
 the Ethernet clock to maintain proper security attributes.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
dts: bindings: ethernet: Add stm32n6 ethernet compat

Add the compatible of the STM32N6 series

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -79,6 +80,10 @@ static const struct device *eth_stm32_phy_dev = DEVICE_PHY_BY_NAME(0);
 #define ETH_MEDIA_INTERFACE_MII		HAL_ETH_MII_MODE
 #define ETH_MEDIA_INTERFACE_RMII	HAL_ETH_RMII_MODE
 
```
Done

fabiobaltieri commented on west.yml:
```diff
@@ -238,7 +238,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: 1d1f81866ccbaa6e84e9960ed763e005d1e45560
+      revision: pull/258/head
```
`7ff81534a75428431ae9d1dd611a4888b4a48219`

marwaiehm-st replied:
Done

marwaiehm-st commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -73,7 +75,12 @@ config ETH_STM32_HAL_PHY_ADDRESS
+config ETH_STM32_HAL_RGMII
+	bool "Use GRMII interface"
+	help
+	  Use the GRMII physical interface instead of RMII and MII.
```
I added the fetch of the interface, but I implemented it in the file 'eth_stm32_hal.c' because I can't read the 'phy-connection-type' from the config file.

maass-hamburg reviewed (changes requested):
in dts/bindings/ethernet/st,stm32-ethernet-common.yaml set this:

```
phy-connection-type:
    default: "rmii"
```

and also remove ``config ETH_STM32_HAL_MII``
and also in board specific Kconfig.defconfig and move it to that dts

maass-hamburg commented:
@marwaiehm-st also pls rebase

erwango commented on soc/st/stm32/stm32n6x/section.ld:
```diff
@@ -0,0 +1,28 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(mac), okay)
+SECTION_DATA_PROLOGUE(eth_stm32, (NOLOAD),)
+{
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(axisram2), okay)
+    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(axisram2)));
+    *(.RxDecripSection)
+    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(axisram2)) + 0xC0);
+    *(.TxDecripSection)
+    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(axisram2)) + 0x4000);
+    *(.eth_stm32_buf)
+} GROUP_DATA_LINK_IN(LINKER_DT_NODE_REGION_NAME(DT_NODELABEL(axisram2)), LINKER_DT_NODE_REGION_NAME(DT_NODELABEL(axisram2)))
+
+#else
+    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(axisram1)));
+    *(.RxDecripSection)
+    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(axisram1)) + 0xC0);
+    *(.TxDecripSection)
+    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(axisram1)) + 0x4000);
+    *(.eth_stm32_buf)
+} GROUP_DATA_LINK_IN(LINKER_DT_NODE_REGION_NAME(DT_NODELABEL(axisram1)), LINKER_DT_NODE_REGION_NAME(DT_NODELABEL(axisram1)))
+#endif
+#endif
```
See https://github.com/zephyrproject-rtos/zephyr/pull/85410

erwango commented on dts/arm/st/n6/stm32n6.dtsi:
```diff
@@ -580,6 +580,25 @@
 		};
+
+		mac: ethernet@58036000 {
+			compatible = "st,stm32n6-ethernet", "st,stm32-ethernet";
```
```suggestion
			compatible = "st,stm32n6-ethernet",  "st,stm32h7-ethernet", "st,stm32-ethernet";
```
This would avoid the need to do things such as :
```
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_ethernet) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)
```

marwaiehm-st replied:
Okay, I will change it.

marwaiehm-st replied:
Done

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -49,6 +49,34 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
+#define MAC_NODE DT_NODELABEL(mac)
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)
+#define STM32_ETH_PHY_MODE(node_id) \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii) ? HAL_ETH_MII_MODE : \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rmii) ? HAL_ETH_RMII_MODE : \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, gmii) ? HAL_ETH_GMII_MODE : \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rgmii) ? HAL_ETH_RGMII_MODE : \
+	HAL_ETH_RMII_MODE))))
+
+#define STM32_ETH_SPEED(node_id) \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii) ? ETH_SPEED_10M : \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rmii) ? ETH_SPEED_100M : \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, gmii) ? ETH_SPEED_1000M : \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rgmii) ? ETH_SPEED_1000M : \
+	ETH_SPEED_100M))))
+#else
+#define STM32_ETH_PHY_MODE(node_id) \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii) ? HAL_ETH_MII_MODE : \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rmii) ? HAL_ETH_RMII_MODE : \
+	HAL_ETH_RMII_MODE))
+
+#define STM32_ETH_SPEED(node_id) \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii) ? ETH_SPEED_10M : \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rmii) ? ETH_SPEED_100M : \
+	ETH_SPEED_100M))
+#endif
+
```
Removing the else case will cause a build failure because 'HAL_ETH_GMII_MODE', 'HAL_ETH_RGMII_MODE', and 'ETH_SPEED_1000M' are only defined on the STM32N6 board.

marwaiehm-st replied:
Done

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1330,8 +1433,7 @@ static struct eth_stm32_hal_dev_data eth0_data = {
 #endif /* !CONFIG_SOC_SERIES_STM32H7X */
-			.MediaInterface = IS_ENABLED(CONFIG_ETH_STM32_HAL_MII) ?
-					  ETH_MEDIA_INTERFACE_MII : ETH_MEDIA_INTERFACE_RMII,
+			// .MediaInterface will be set at runtime
```
done

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1024,8 +1127,8 @@ static int eth_initialize(const struct device *dev)
-	mac_config.Speed = IS_ENABLED(CONFIG_ETH_STM32_SPEED_10M) ?
-				 ETH_SPEED_10M : ETH_SPEED_100M;
+	mac_config.Speed = get_mac_speed();
+
```
done

erwango commented on dts/bindings/ethernet/st,stm32-ethernet-common.yaml:
```diff
@@ -18,3 +18,5 @@ properties:
   pinctrl-names:
     required: true
+  phy-connection-type:
+    default: "rmii"
```
See https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules, but I'm not sure default really matches constraints here.

maass-hamburg replied:
it is done the same as in dts/bindings/ethernet/atmel,gmac-common.yaml. 
By the way, this helps keeping the compatibility, as that was the default for the Kconfigs (you had to select a Kconfig if you wanted something else than rmii).

erwango commented:
> This does not seem right, adding a compatible to reduce code. That is more like a missuse of the DT.
Maybe use the Kconfig CONFIG_SOC_SERIES_STM32N6X for filtering in the driver. this would also eliminale having to have a seperate dt compatible.

@maass-hamburg This is perfectly fine. Here is the specification description of "compatible" property:
> The compatible property value consists of one or more strings that define the specific programming model for
the device. This list of strings should be used by a client program for device driver selection. The property
value consists of a concatenated list of null terminated strings, from most specific to most general. They allow
a device to express its compatibility with a family of similar devices, potentially allowing a single device driver
to match against several devices.

So here most specific is "st,stm32n6-ethernet" while most general is "st,stm32-ethernet". 
"st,stm32-ethernet" is used to select the driver that will be used and "st,stm32n6-ethernet" allows to make some customization on top of it. This model is used in other drivers as well.
Filtering using `CONFIG_SOC_SERIES_STM32N6X` as you propose will only make driver harder to read and maintain. This is where we're coming from and use of compatible make drivers much easier to maintain (as for instance you don't need to update the driver the next time a compatible series should be added. You just need to provide the right compatible in the .dtsi).

maass-hamburg commented:
> > This does not seem right, adding a compatible to reduce code. That is more like a missuse of the DT.
> > Maybe use the Kconfig CONFIG_SOC_SERIES_STM32N6X for filtering in the driver. this would also eliminale having to have a seperate dt compatible.
> 
> @maass-hamburg This is perfectly fine. Here is the specification description of "compatible" property:
> 
> > The compatible property value consists of one or more strings that define the specific programming model for
> > the device. This list of strings should be used by a client program for device driver selection. The property
> > value consists of a concatenated list of null terminated strings, from most specific to most general. They allow
> > a device to express its compatibility with a family of similar devices, potentially allowing a single device driver
> > to match against several devices.
> 
> So here most specific is "st,stm32n6-ethernet" while most general is "st,stm32-ethernet". "st,stm32-ethernet" is used to select the driver that will be used and "st,stm32n6-ethernet" allows to make some customization on top of it. This model is used in other drivers as well. Filtering using `CONFIG_SOC_SERIES_STM32N6X` as you propose will only make driver harder to read and maintain. This is where we're coming from and use of compatible make drivers much easier to maintain (as for instance you don't need to update the driver the next time a compatible series should be added. You just need to provide the right compatible in the .dtsi).

I was referring the suggestion to have  st,stm32n6-ethernet" and  "st,stm32h7-ethernet" in the compatible. the generic "st,stm32-ethernet". was fine.

erwango commented:
> I was referring the suggestion to have st,stm32n6-ethernet" and "st,stm32h7-ethernet" in the compatible. the generic "st,stm32-ethernet". was fine.

Ok, but I still think it is correct, `st,stm32n6-ethernet` being seen as a specific case of `st,stm32h7-ethernet`.
And as could be seen this allows to get rid of all the `#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_ethernet) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)`

maass-hamburg commented:
> > I was referring the suggestion to have st,stm32n6-ethernet" and "st,stm32h7-ethernet" in the compatible. the generic "st,stm32-ethernet". was fine.
> 
> Ok, but I still think it is correct, `st,stm32n6-ethernet` being seen as a specific case of `st,stm32h7-ethernet`. And as could be seen this allows to get rid of all the `#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_ethernet) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)`

would the st,stm32n6 work with just ``st,stm32h7-ethernet`` and ``st,stm32-ethernet`` as the compatible?

erwango commented:
> would the st,stm32n6 work (even limited) with just st,stm32h7-ethernet and st,stm32-ethernet as the compatible?

You can replace `#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)` by `#ifdef CONFIG_SOC_SERIES_STM32N6X`. But then, when adding support for a next series  fullycompatible with N6 IP, you'll have to do:
`#ifdef CONFIG_SOC_SERIES_STM32N6X || CONFIG_SOC_SERIES_STM32YZX`.
Using compat method, you'll just need to provide the right compatibles in new series .dtsi and you'll be done. No need to touch the driver.

maass-hamburg commented:
> > would the st,stm32n6 work (even limited) with just st,stm32h7-ethernet and st,stm32-ethernet as the compatible?
> 
> You can replace `#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)` by `#ifdef CONFIG_SOC_SERIES_STM32N6X`. But then, when adding support for a next series fullycompatible with N6 IP, you'll have to do: `#ifdef CONFIG_SOC_SERIES_STM32N6X || CONFIG_SOC_SERIES_STM32YZX`. Using compat method, you'll just need to provide the right compatibles in new series .dtsi and you'll be done. No need to touch the driver.

I understand you, but that was not the point. A compatible should not depend on another compatible being present. The second and later compatibles are meant as a fallback, if there is no diver available for the first compatible available. I don't like having a chain of compatibles, that all depend on the next.-

erwango commented:
> I understand you, but that was not the point. A compatible should not depend on another compatible being present. The second and later compatibles are meant as a fallback, if there is no diver available for the first compatible available. I don't like having a chain of compatibles, that all depend on the next.-

This fallback mechanism is valid for Linux but hardly applies in Zephyr, right ?
What other solution do you propose that avoids to update driver code each time a new series with compatible IP should be supported by the driver code (which is the final vocation of "compatible" in the end) ?

maass-hamburg commented:
> > I understand you, but that was not the point. A compatible should not depend on another compatible being present. The second and later compatibles are meant as a fallback, if there is no diver available for the first compatible available. I don't like having a chain of compatibles, that all depend on the next.-
> 
> This fallback mechanism is valid for Linux but hardly applies in Zephyr, right ? What other solution do you propose that avoids to update driver code each time a new series with compatible IP should be supported by the driver code (which is the final vocation of "compatible" in the end) ?

If it is compatible with a former compatible, use that. If there are new features, that requires the creation of a new compatible, create it and then make the appropriate changes to the driver and add ``|| DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)`` where it belongs. Maybe create a define for ``DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_ethernet) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)`` at be beginning of the file to shorten it.

maass-hamburg commented:
Are there cases where users have to set the compatible themself for Ethernet? Or do they always have to include the dtsi?

If they never have to set it themself (in out of tree boards f.e.), I still don't like it, but can tolerate it.

erwango commented:
> If it is compatible with a former compatible, use that. If there are new features, that requires the creation of a new compatible, create it and then make the appropriate changes to the driver and add || DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet) where it belongs.

That is exactly what is done in the PR currently:
>  If it is compatible with a former compatible, use that.

N6 Eth IP makes use of existing STM32 driver (requires `st,stm32-ethernet`) and is a H7 compatible variant of it, which hence requires use of `st,stm32h7-ethernet` on top. 

> If there are new features, that requires the creation of a new compatible, create it and then make the appropriate changes to the driver and add || DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet) where it belongs.

Done as well. Hence the introduction of `st,stm32n6-ethernet`.

Finally, following these steps, we end up with `compatible = st,stm32n6-ethernet, st,stm32h7-ethernet, st,stm32-ethernet`.

maass-hamburg commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -9,11 +9,12 @@ menuconfig ETH_STM32_HAL
 	select ETH_DSA_SUPPORT
 	select PINCTRL
-	select MDIO if SOC_SERIES_STM32H5X || SOC_SERIES_STM32H7X
+	select MDIO if SOC_SERIES_STM32H5X || SOC_SERIES_STM32H7X || SOC_SERIES_STM32N6X
```
here the mdio subsystem is being enabled. the correct driver for mdio is automatically being enabled by ``DT_HAS_ST_STM32_MDIO_ENABLED``

maass-hamburg commented:
@marwaiehm-st this should be rebased

maass-hamburg commented:
@marwaiehm-st RGMII support is now missing, was that intended?

marwaiehm-st commented:
> @marwaiehm-st RGMII support is now missing, was that intended?

The RGMII is used on the board STM32N6570_DK, and the support for the STM32N6570_DK will be added in a separate PR.

maass-hamburg commented:
Please add the ``st,stm32n6-ethernet.yaml`` file.
